feat: LEAP-1020: Relation side panel improvements

Precious-Macaulay opened the pull request:
See #5446
<!-- feat: Relations side panel improvements -->
## Overview
This PR implements improvements to the relations list based on the accepted [prd: Relations list improvements]
See #5689 #5447 #5446 


https://github.com/HumanSignal/label-studio/assets/87583799/ec5276b1-b940-4d55-ac54-546380a6ef0b




## Changes

### Relations list ordering
- Added ordering options to relations list sidepanel 
- Ordering by creation time (ascending or descending)
- Ordering preference persists between sessions

### Relations visibility toggle  
- Added toggle icon next to "Relations" label
- Toggle hides all relations when one or more are visible
- Toggle shows all relations when all are hidden
- Visibility state persists between sessions

## Testing
- Added unit tests for ordering and visibility functionality
- Updated documentation with usage instructions

## PR Requirements 

- [x] Commit message(s) and PR title follows the format 
- [x] Tests for the changes have been added/updated
- [x] Docs have been added/updated  
- [x] Best efforts for concise/coherent code

#### Areas Impacted
- [ ] Product design
- [ ] Backend (Database) 
- [ ] Backend (API)
- [x] Frontend

### Change Details
This is a new feature implementing relations list ordering and visibility toggling based on the provided requirements.

No libraries were added or updated.

This change should have a positive performance impact by allowing users to easily focus on relevant relations.

This change does not impact security.

Alternatives were manually hiding relations and scrolling down which is tedious.

No feature flags were used.

### Breaking Change
- [ ] Yes, and covered entirely by feature flags
- [ ] Yes, and covered partially by feature flags  
- [x] No
- [ ] Not sure

### Testing 
- [ ] e2e
- [x] integration
- [ ] unit

### Affected Domains
- Relations Details Panel

## Acceptance Criteria
- [x] Adds options to order relations list by creation time (ascending and descending)
- [x] Persists ordering preference between sessions 
- [x] Integration test for ordering functionality

- [x] Adds a toggle icon next to "Relations" label
- [x] Toggle hides all relations when one or more are visible
- [x] Toggle shows all relations when all are hidden
- [x] Persists visibility state between sessions 
- [x] Integration tests for visibility functionality

Closes #5689 #5447 #5446

Precious-Macaulay committed:
feat: Relation side panel improvements (#5689)

Precious-Macaulay commented:
@deppp @hlomzik @jombooth do you mind checking the code?
I have improved the code and added integration test ( #5694 )

hlomzik commented:
@Precious-Macaulay thank you, this PR looks really good and well-thought! I'll check it live soon and we have to run it through our internal processes. One small comment right now — new files have too big indent, we use 2 spaces everywhere, but that's just a linting issue.

Precious-Macaulay commented:
Thanks for the feedback! I'll make sure to adjust the indentations accordingly to match the 2-space standard.

makseq commented:
There is our internal ticket for this PR: https://humansignal.atlassian.net/browse/LEAP-1020

Precious-Macaulay committed:
fix: LEAP-1020: fixed linting issues and all failed test

Precious-Macaulay commented:
> @Precious-Macaulay thank you, this PR looks really good and well-thought! I'll check it live soon and we have to run it through our internal processes. One small comment right now — new files have too big indent, we use 2 spaces everywhere, but that's just a linting issue.

I committed a fix for this

Precious-Macaulay commented:
> There is our internal ticket for this PR: https://humansignal.atlassian.net/browse/LEAP-1020

Alright

Precious-Macaulay commented:
@hlomzik any update on this

bmartel commented on web/libs/editor/src/components/SidePanels/DetailsPanel/Relations.tsx:
```diff
@@ -10,9 +10,13 @@ import { Select } from 'antd';
 import './Relations.styl';
 
 const RealtionsComponent: FC<any> = ({ relationStore }) => {
+  const relations = useCallback(() => {
+    return relationStore.orderedRelations;
+  }, [relationStore])();
```
```suggestion
  const relations = relationStore.orderedRelations;
```
I'm not sure what the use case was for using a useCallback like a useMemo, but from what I can see this would be more likely the scenario that would be expected here is to allow the reactive property to be passed through directly.

bmartel commented on web/libs/frontend-test/src/helpers/LSF/Relations.ts:
```diff
@@ -1,15 +1,53 @@
   toggleCreation() {
     cy.get('.lsf-region-actions__group_align_left > :nth-child(1) > .lsf-button__icon').click();
   },
-  toggleCreationWithHotkey() {
-    // hotkey is alt + r
-    cy.get('body').type('{alt}r');
-  },
```
Why was this helper method removed? I just did a quick search and it is still used elsewhere in the codebase.

Precious-Macaulay replied:
Apologies for the oversight. I've restored the accidentally deleted helper method.

bmartel commented on web/libs/editor/tests/integration/e2e/relations/order.cy.ts:
```diff
@@ -0,0 +1,175 @@
+  data: { image },
+};
+
+describe('Relations- Sort Order all relations', () => {
```
```suggestion
describe('Relations: Sort Order all relations', () => {
```

bmartel commented on web/libs/editor/tests/integration/e2e/relations/hide-all.cy.ts:
```diff
@@ -0,0 +1,207 @@
+  data: { image },
+};
+
+describe('Relations- Hide/Show all relations', () => {
```
```suggestion
describe('Relations: Hide/Show all relations', () => {
```

Precious-Macaulay committed:
Apply suggestions from code review

Co-authored-by: bmartel <brandonmartel@gmail.com>

Precious-Macaulay committed:
Restored Deleted Helper Method

Precious-Macaulay committed:
removed unnecessary usecallback

Apologies for the oversight. The use case was intended to prevent unnecessary re-renders, which is typically achieved using `useMemo` rather than `useCallback`. However, considering it is a reactive property, employing either may be superfluous.

Co-authored-by: bmartel <brandonmartel@gmail.com>

Precious-Macaulay committed:
Merge branch 'develop' into develop

Precious-Macaulay committed:
minor fix

bmartel commented:
Thanks for making those changes @Precious-Macaulay! I noticed a few discrepancies in formatting when going over the code again. Would you be able to run the formatter/linter locally and push the changes up?

```bash
cd web && yarn run biome check --apply .
```

Precious-Macaulay committed:
Fix formatting issues

Precious-Macaulay reviewed (commented):
> Thanks for making those changes @Precious-Macaulay! I noticed a few discrepancies in formatting when going over the code again. Would you be able to run the formatter/linter locally and push the changes up?
> 
> ```bash
> cd web && yarn run biome check --apply .
> ```


Yes, I have fixed the formatting and resolved the linting issues.

Precious-Macaulay committed:
Fix relation visibility toggle in integration tests and helper method in frontend-test

Precious-Macaulay committed:
Merge branch 'HumanSignal:develop' into develop

Precious-Macaulay commented:
@makseq @bmartel @hlomzik any update on this PR

sajarin commented:
hey @Precious-Macaulay, we are closing to getting this PR merged, thanks for being patient! 🙏

Precious-Macaulay commented:
> hey @Precious-Macaulay, we are closing to getting this PR merged, thanks for being patient! 🙏

Thank you for the update! I'm excited to see it get merged. 😊

hlomzik commented:
Thank you, @Precious-Macaulay, it's there! Really good job!
